File: SmartInventoryClient/src/app/services/admin/category-service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Category } from '../../models/category';

@Injectable({
  providedIn: 'root',
})
export class CategoryService {

  private api = 'http://localhost:5122/api/categories';

  constructor(private http: HttpClient) {}

  // GET all categories
  getAll() {
    return this.http.get<Category[]>(this.api);
  }

  // ADD category
  add(name: string) {
    return this.http.post(this.api, { name });
  }

  // UPDATE category
  update(id: number, data: Partial<Category>) {
    return this.http.put(`${this.api}/${id}`, data);
  }

  // DELETE category
  delete(id: number) {
    return this.http.delete(`${this.api}/${id}`);
  }
}

File: SmartInventorySystem.Api/SmartInventorySystem.Api/Services/Categories/CategoryService.cs
using Microsoft.EntityFrameworkCore;
using SmartInventorySystem.Api.Data;
using SmartInventorySystem.Api.Models.DTOs;
using SmartInventorySystem.Api.Models.Entities;

namespace SmartInventorySystem.Api.Services.Categories
{
    public class CategoryService(AppDbContext context) : ICategoryService
    {
        private readonly AppDbContext _context = context;

        // ================= GET ALL =================
        public List<Category> GetAll()
        {
            return _context.Categories.ToList();
        }

        // ================= CREATE =================
        public Category Create(CategoryDto dto)
        {
            var category = new Category
            {
                Name = dto.Name
            };

            _context.Categories.Add(category);
            _context.SaveChanges();

            return category;
        }

        // ================= UPDATE =================
        public Category Update(int id, CategoryDto dto)
        {
            var category = _context.Categories.Find(id);

            if (category == null)
                throw new KeyNotFoundException("Category not found");

            category.Name = dto.Name;
            _context.SaveChanges();

            return category;
        }

        // ================= DELETE =================
        public void Delete(int id)
        {
            var category = _context.Categories
                .Include(c => c.Products)
                .FirstOrDefault(c => c.Id == id);

            if (category == null)
                throw new KeyNotFoundException("Category not found");

            if (category.Products.Any())
                throw new InvalidOperationException(
                    "Cannot delete category with assigned products"
                );

            _context.Categories.Remove(category);
            _context.SaveChanges();
        }
    }
}
